static/app/utils/useRequest.tsx

```diff
@@ -0,0 +1,252 @@
+  } = {};
+
+  if (search) {
+    query.query = `${query.query ?? ''} ${search}`.trim();
+  }
+
+  const isSameSearch = lastSearch === search || (!lastSearch && !search);
+
+  if (isSameSearch && cursor) {
+    query.cursor = cursor;
+  }
+
+  if (limit !== undefined) {
+    query.per_page = limit;
+  }
+
+  let hasMore: null | boolean = false;
+  let nextCursor: null | string = null;
+  const [data, , resp] = await api.requestPromise(url, {
+    includeAllArgs: true,
+    query,
+  });
+
+  const pageLinks = resp?.getResponseHeader('Link');
+  if (pageLinks) {
+    const paginationObject = parseLinkHeader(pageLinks);
+    hasMore = paginationObject?.next?.results || paginationObject?.previous?.results;
+    nextCursor = paginationObject?.next?.cursor;
+  }
+
+  return {results: data, hasMore, nextCursor};
+}
+
+function useRequest(
+  url: string,
+  {limit}: Options = {},
+  onSuccessCb = _ => {},
+  onErrorCb = _ => {}
+) {
+  const api = useApi();
+  const {organization} = useLegacyStore(OrganizationStore);
+
+  const orgId = organization?.slug;
+
+  const [state, setState] = useState<State>({
+    data: null,
+    loading: false,
+    hasMore: null,
+    lastSearch: null,
+    nextCursor: null,
+    error: null,
+  });
+
+  async function get(cursor: string | null = null) {
+    setState({...state, loading: true});
+    try {
+      const {results, hasMore, nextCursor} = await fetch(api, url, {
+        limit,
+        cursor,
+      });
+      onSuccessCb(results);
+      setState({
+        ...state,
+        data: results,
+        hasMore,
+        loading: false,
+        nextCursor,
+      });
+    } catch (err) {
+      Sentry.captureException(err);
+      setState({...state, loading: false, error: err});
+      onErrorCb(err);
+    }
+  }
+
+  async function handleSearch(search: string) {
+    const {lastSearch} = state;
+    const cursor = state.nextCursor;
+
+    if (search === '') {
+      return;
+    }
+
+    if (orgId === undefined) {
+      // eslint-disable-next-line no-console
+      console.error('Cannot use useRequest.onSearch without an organization in context');
+      return;
+    }
+
+    setState({...state, loading: true});
+
+    try {
+      api.clear();
+      const {results, hasMore, nextCursor} = await fetch(api, url, {
+        search,
+        limit,
+        lastSearch,
+        cursor,
+      });
+
+      onSuccessCb(results);
+
+      setState({
+        ...state,
+        data: results,
+        hasMore,
+        loading: false,
+        lastSearch: search,
+        nextCursor,
+      });
+    } catch (err) {
+      Sentry.captureException(err);
+      setState({...state, loading: false, error: err});
+      onErrorCb(err);
+    }
+  }
+
+  async function createOrUpdate(data: any, method: 'POST' | 'PUT' = 'POST') {
+    setState({...state, loading: true});
+    try {
+      const [results] = await api.requestPromise(url, {
+        method,
+        data,
+      });
+      onSuccessCb(results);
```

Comment:
If you returned a Promise callers could use `await` or `.then()` which gives more flexibility than a callback would. The `onErrorCb` could be a reject on the promise.